### PR TITLE
[MINOR] Fix chunked down-conversion behavior when no valid batch exists after conversion

### PR DESCRIPTION
**Pull request by dhruvilshah3**
We might decide to drop certain message batches during down-conversion because older clients might not be able to interpret them. One such example is control batches which are typically removed by the broker if down-conversion to V0 or V1 is required. This patch makes sure the chunked down-conversion implementation is able to handle such cases.

Credit to @omkreddy for reporting this issue, providing logs and a reproducible test case.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/test/java/org/apache/kafka/common/record/LazyDownConversionRecordsTest.java [hunk 1]
```diff
@@ -123,50 +55,138 @@ public void doTestConversion() throws IOException {
...
-
-            convertedRecords.close();
-            channel.close();
+            MemoryRecords convertedRecords = convertRecords(memoryRecords, toMagic, toWrite);
+            verifyDownConvertedRecords(recordsBeingConverted, offsetsOfRecords, convertedRecords, compressionType, toMagic);
         }
     }
 
-    private String utf8(ByteBuffer buffer) {
+    private static String utf8(ByteBuffer buffer) {
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/common/record/LazyDownConversionRecordsTest.java]: nit: why don't we add this to `Utils`?

File: clients/src/main/java/org/apache/kafka/common/record/LazyDownConversionRecords.java [hunk 2]
```diff
@@ -148,21 +149,24 @@ protected ConvertedRecords makeNext() {
...
+                // Figure out batches we should down-convert based on the size constraints
+                while (batchIterator.hasNext() &&
+                        (isFirstBatch || (batchIterator.peek().sizeInBytes() + sizeSoFar) <= maximumReadSize)) {
+                    RecordBatch currentBatch = batchIterator.next();
+                    batches.add(currentBatch);
+                    sizeSoFar += currentBatch.sizeInBytes();
+                    isFirstBatch = false;
+                }
+                ConvertedRecords records = RecordsUtil.downConvert(batches, toMagic, firstOffset, time);
+                if (records.records().sizeInBytes() > 0)
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/record/LazyDownConversionRecords.java]: It's probably worth a comment here explaining which situations would result in 0 down-converted bytes.

File: clients/src/main/java/org/apache/kafka/common/record/LazyDownConversionRecordsSend.java [hunk 3]
```diff
@@ -49,39 +49,28 @@ public LazyDownConversionRecordsSend(String destination, LazyDownConversionRecor
...
-
-                int sizeOfFirstConvertedBatch = convertedRecords.batchIterator().next().sizeInBytes();
-                if (previouslyWritten == 0 && sizeOfFirstConvertedBatch > size())
-                    throw new EOFException("Unable to send first batch completely." +
-                            " maximum_size: " + size() +
-                            " converted_records_size: " + sizeOfFirstConvertedBatch);
-
                 recordConversionStats.add(recordsAndStats.recordConversionStats());
-                log.debug("Got lazy converted records for partition {} with length={}", topicPartition(), convertedRecords.sizeInBytes());
+                log.debug("Got converted records for partition {} with length={}", topicPartition(), convertedRecords.sizeInBytes());
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/record/LazyDownConversionRecordsSend.java]: nit: "Got" is not very descriptive. Maybe the message could be "Down-converted records partition ..."?

File: clients/src/main/java/org/apache/kafka/common/record/LazyDownConversionRecordsSend.java [hunk 4]
```diff
@@ -49,39 +49,28 @@ public LazyDownConversionRecordsSend(String destination, LazyDownConversionRecor
...
                 //      ...
-                log.debug("Constructing fake message batch for partition {} for remaining length={}", topicPartition(), remaining);
-                ByteBuffer fakeMessageBatch = ByteBuffer.allocate(Math.max(Records.LOG_OVERHEAD, Math.min(remaining + 1, MAX_READ_SIZE)));
-                fakeMessageBatch.putLong(-1L);
-                fakeMessageBatch.putInt(remaining + 1);
-                convertedRecords = MemoryRecords.readableRecords(fakeMessageBatch);
+                ByteBuffer overflowMessageBatch = ByteBuffer.allocate(
+                        Math.max(MIN_OVERFLOW_MESSAGE_LENGTH, Math.min(remaining + 1, MAX_READ_SIZE)));
+                overflowMessageBatch.putLong(-1L);
+                overflowMessageBatch.putInt(Math.max(remaining + 1, DefaultRecordBatch.RECORD_BATCH_OVERHEAD));
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/record/LazyDownConversionRecordsSend.java]: It's not clear to me why we need the `max` here. Since we just need something that overflows, I'm wondering why we don't use something like `Integer.MAX_VALUE`?

File: clients/src/test/java/org/apache/kafka/common/record/LazyDownConversionRecordsTest.java [hunk 5]
```diff
@@ -41,80 +41,12 @@
...
-
-        builder = MemoryRecords.builder(buffer, RecordBatch.CURRENT_MAGIC_VALUE, compressionType, TimestampType.CREATE_TIME,
-                0L);
-        for (int i = 6; i < 10; i++)
-            builder.appendWithOffset(offsets.get(i), records.get(i));
-        builder.close();
-
-        buffer.flip();
 
+    public static MemoryRecords convertRecords(MemoryRecords recordsToConvert, byte toMagic, int toWrite) throws IOException {
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/common/record/LazyDownConversionRecordsTest.java]: nit: could be `private`. For what it's worth, I tend to prefer having these helper methods at the bottom of the file so that you don't have to hunt for the test cases, but it's a bit subjective.

File: clients/src/test/java/org/apache/kafka/common/record/LazyDownConversionRecordsTest.java [hunk 6]
```diff
@@ -123,50 +55,138 @@ public void doTestConversion() throws IOException {
...
+        private final byte toMagic;
+        private final DownConversionTest test;
+
+        public ParameterizedConversionTest(CompressionType compressionType, byte toMagic, DownConversionTest test) {
+            this.compressionType = compressionType;
+            this.toMagic = toMagic;
+            this.test = test;
+        }
+
+        enum DownConversionTest {
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/common/record/LazyDownConversionRecordsTest.java]: I think it would be more straightforward to have two separate test cases which call `doTestConversion` with a boolean parameter to indicate whether overflow should be tested.

File: clients/src/test/java/org/apache/kafka/common/record/LazyDownConversionRecordsTest.java [hunk 7]
```diff
@@ -123,50 +55,138 @@ public void doTestConversion() throws IOException {
...
+
+            builder = MemoryRecords.builder(buffer, RecordBatch.CURRENT_MAGIC_VALUE, compressionType, TimestampType.CREATE_TIME,
+                    0L);
+            for (int i = 6; i < 10; i++)
+                builder.appendWithOffset(offsets.get(i), records.get(i));
+            builder.close();
+
+            buffer.flip();
+
+            MemoryRecords memoryRecords = MemoryRecords.readableRecords(buffer);
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/common/record/LazyDownConversionRecordsTest.java]: nit: maybe we can think of a more descriptive name. Maybe "recordsToConvert"?

**Review by hachikuji — COMMENTED** [review]
Thanks, the fix looks good. Left a few small comments.

File: clients/src/test/java/org/apache/kafka/common/record/FileRecordsTest.java [hunk 8]
```diff
@@ -431,7 +431,7 @@ private void doTestConversion(CompressionType compressionType, byte toMagic) thr
     }
 
     private String utf8(ByteBuffer buffer) {
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/common/record/FileRecordsTest.java]: This method no longer seems worthwhile. Same in `MemoryRecordsBuilderTest`.

File: docs/upgrade.html [hunk 9]
```diff
@@ -119,6 +119,13 @@ <h5><a id="upgrade_200_notable" href="#upgrade_200_notable">Notable changes in 2
     </li>
     <li><a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-290%3A+Support+for+Prefixed+ACLs">KIP-290</a> adds the ability
         to define ACLs on prefixed resources, e.g. any topic starting with 'foo'.</li>
+    <li><a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-283%3A+Efficient+Memory+Usage+for+Down-Conversion">KIP-283</a> improves message down-conversion
+        handling on Kafka broker, which has typically been a memory-intensive operation. The KIP adds new topic and broker configurations <code>message.downconversion.enable</code>
```
Comment (hachikuji) [inline docs/upgrade.html]: I'd suggest moving the mention of the config to the end. The improvement in memory utilization is the core feature here.

**Review by hachikuji — COMMENTED** [review]
Thanks, just a couple additional comments.

**Review by hachikuji — APPROVED** [review]
LGTM, thanks for the patch!
